File: ios/RailwayMapCore/Sources/RailwayMapCore/Prefs.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

import Foundation

/// Camera position as saved between launches.
public struct SavedCamera: Hashable, Sendable {
    public var lat: Double
    public var lon: Double
    public var zoom: Double

    public init(lat: Double, lon: Double, zoom: Double) {
        self.lat = lat
        self.lon = lon
        self.zoom = zoom
    }
}

/// Tiny wrapper around UserDefaults for the handful of things we persist; same keys as on Android.
public final class Prefs: @unchecked Sendable {
    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private func string(_ key: String, _ fallback: String) -> String {
        defaults.string(forKey: key) ?? fallback
    }

    private func bool(_ key: String, _ fallback: Bool) -> Bool {
        defaults.object(forKey: key) == nil ? fallback : defaults.bool(forKey: key)
    }

    public var modeId: String {
        get { string("mode", "standard") }
        set { defaults.set(newValue, forKey: "mode") }
    }

    public var showConstruction: Bool {
        get { bool("showConstruction", true) }
        set { defaults.set(newValue, forKey: "showConstruction") }
    }

    public var showProposed: Bool {
        get { bool("showProposed", true) }
        set { defaults.set(newValue, forKey: "showProposed") }
    }

    public var showAbandoned: Bool {
        get { bool("showAbandoned", false) }
        set { defaults.set(newValue, forKey: "showAbandoned") }
    }

    public var showRazed: Bool {
        get { bool("showRazed", false) }
        set { defaults.set(newValue, forKey: "showRazed") }
    }

    public var electrificationLine: String {
        get { string("electrificationLine", "voltageFrequency") }
        set { defaults.set(newValue, forKey: "electrificationLine") }
    }

    public var trackLine: String {
        get { string("trackLine", "gauge") }
        set { defaults.set(newValue, forKey: "trackLine") }
    }

    public var stationLowZoomLabel: String {
        get { string("stationLowZoomLabel", "label") }
        set { defaults.set(newValue, forKey: "stationLowZoomLabel") }
    }

    /// Last camera position; nil when never saved.
    public var camera: SavedCamera? {
        get {
            guard defaults.object(forKey: "cam_lat") != nil else {
                return nil
            }
            return SavedCamera(
                lat: defaults.double(forKey: "cam_lat"),
                lon: defaults.double(forKey: "cam_lon"),
                zoom: defaults.double(forKey: "cam_zoom")
            )
        }
        set {
            if let camera = newValue {
                defaults.set(camera.lat, forKey: "cam_lat")
                defaults.set(camera.lon, forKey: "cam_lon")
                defaults.set(camera.zoom, forKey: "cam_zoom")
            } else {
                for key in ["cam_lat", "cam_lon", "cam_zoom"] {
                    defaults.removeObject(forKey: key)
                }
            }
        }
    }
}

File: android/app/src/main/java/com/offlinerailmap/android/data/Prefs.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Joel Haasnoot

package com.offlinerailmap.android.data

import android.content.Context
import android.content.SharedPreferences

/** Tiny wrapper around SharedPreferences for the handful of things we persist. */
object Prefs {
    private lateinit var prefs: SharedPreferences

    fun init(context: Context) {
        prefs = context.applicationContext.getSharedPreferences("offline_railway_map", Context.MODE_PRIVATE)
    }

    var modeId: String
        get() = prefs.getString("mode", "standard") ?: "standard"
        set(value) = prefs.edit().putString("mode", value).apply()

    var showConstruction: Boolean
        get() = prefs.getBoolean("showConstruction", true)
        set(v) = prefs.edit().putBoolean("showConstruction", v).apply()

    var showProposed: Boolean
        get() = prefs.getBoolean("showProposed", false)
        set(v) = prefs.edit().putBoolean("showProposed", v).apply()

    var showAbandoned: Boolean
        get() = prefs.getBoolean("showAbandoned", false)
        set(v) = prefs.edit().putBoolean("showAbandoned", v).apply()

    var showRazed: Boolean
        get() = prefs.getBoolean("showRazed", false)
        set(v) = prefs.edit().putBoolean("showRazed", v).apply()

    var electrificationLine: String
        get() = prefs.getString("electrificationLine", "voltageFrequency") ?: "voltageFrequency"
        set(v) = prefs.edit().putString("electrificationLine", v).apply()

    var trackLine: String
        get() = prefs.getString("trackLine", "gauge") ?: "gauge"
        set(v) = prefs.edit().putString("trackLine", v).apply()

    var stationLowZoomLabel: String
        get() = prefs.getString("stationLowZoomLabel", "label") ?: "label"
        set(v) = prefs.edit().putString("stationLowZoomLabel", v).apply()

    var darkTheme: Boolean
        get() = prefs.getBoolean("darkTheme", false)
        set(v) = prefs.edit().putBoolean("darkTheme", v).apply()

    /** Last camera position as lat, lon, zoom; null when never saved. */
    var camera: Triple<Double, Double, Double>?
        get() {
            if (!prefs.contains("cam_lat")) {
                return null
            }
            return Triple(
                prefs.getFloat("cam_lat", 0f).toDouble(),
                prefs.getFloat("cam_lon", 0f).toDouble(),
                prefs.getFloat("cam_zoom", 0f).toDouble(),
            )
        }
        set(v) {
            if (v == null) {
                prefs.edit().remove("cam_lat").remove("cam_lon").remove("cam_zoom").apply()
            } else {
                prefs.edit()
                    .putFloat("cam_lat", v.first.toFloat())
                    .putFloat("cam_lon", v.second.toFloat())
                    .putFloat("cam_zoom", v.third.toFloat())
                    .apply()
            }
        }
}
